src/Build.UnitTests/Utilities_Tests.cs

```diff
@@ -63,6 +60,23 @@ public void GetTextFromTextNodeWithXmlComment7()
             // Should get XML; note space after x added
             Assert.Equal("<!-- bar; baz; --><!-- bar --><x />", xmlContents);
         }
+
+        [Fact]
+        public void HandshakesDiffer()
+        {
+            int numHandshakeOptions = (int)Math.Pow(2, Enum.GetNames(HandshakeOptions.None.GetType()).Length - 1);
+            Dictionary<long, int> handshakes = new Dictionary<long, int>();
+            for (int i = 0; i < numHandshakeOptions; i++)
+            {
+                long nextKey = CommunicationsUtilities.GetHostHandshake((HandshakeOptions)i);
+                if (handshakes.TryGetValue(nextKey, out int collision))
+                {
+                    CommunicationsUtilities.Trace("There was a collision between {0} and {1}.", collision, i);
+                }
+                handshakes.Add(nextKey, i);
+            }
+            handshakes.Count.ShouldBe(numHandshakeOptions, "two or more combinations of handshake options hashed to the same value");
```

Comment:
Collect into an array so you can use the helper function with a nice assert?

```suggestion
            handshakes.ShouldBeUnique("two or more combinations of handshake options hashed to the same value");
```

Reply:
If I leave the current tracing code, making this a real check would mean duplicating that work, but the current tracing is cleaner in my opinion for reasons outlined below. The real purpose of this check is just to make sure the test fails if there are duplicates so someone knows to look into it.

The various handshakeoptions are obviously going to be unique, including if I made the items in the array tuples, so I'd have to make it an array of the handshakes, and those are hard to understand. You could still reverse-engineer which pair(s) are colliding by looking at the index of the two that are the same, but beyond identifying which pairs are colliding when there can be very similar pairs of handshakes, you'd have to count to the indices to see which handshakeoptions they represent, and that's already dicey with as many options as there are and will get worse if we add more. I think it's doable, but it would make it harder to use.